test/boost/type_json_test.cc

```diff
@@ -0,0 +1,183 @@
+    }
+
+    double next_abs_greater() const {
+        auto result = *this;
+        if (mantissa_bits().all()) {
+            // "carry" a power of 2
+            result.set_exponent(exponent() + 1);
+            result.set_mantissa(0);
+        } else {
+            result.set_mantissa(mantissa() + 1);
+        }
+        return double(result);
+    }
+
+    double next_abs_lesser() const {
+        auto result = *this;
+        if (mantissa_bits().none()) {
+            // "borrow" a power of 2
+            result.set_exponent(exponent() - 1);
+            result.set_mantissa(std::numeric_limits<uint64_t>::max());
+        } else {
+            result.set_mantissa(mantissa() - 1);
+        }
+        return double(result);
+    }
+
+public:
+    // construct a 64-bit float point number which is closest to
+    // the specified "n" by casting it to double
+    static IEEE754_double closest(int64_t n) {
+        static_assert(std::numeric_limits<double>::round_style == std::float_round_style::round_to_nearest);
+        auto fp64 = static_cast<double>(n);
+        return IEEE754_double(std::bit_cast<uint64_t>(fp64));
+    }
+
+    explicit operator double() const {
+        return std::bit_cast<double>(_bits.to_ulong());
+    }
+
+    // returns the maximum number less than "n"
+    double previous() const {
+        if (is_negative()) {
+            return next_abs_greater();
+        } else {
+            return next_abs_lesser();
+        }
+    }
+
+    double next() const {
+        if (is_negative()) {
+            return next_abs_lesser();
+        } else {
+            return next_abs_greater();
+        }
+    }
+};
+
+// should reject the numbers not representable by int64_t by throwing
+// marshal_exception.
+BOOST_AUTO_TEST_CASE(test_from_json_object_to_long_type) {
+  // double, which is IEEE-754 64 float point. despite that it can cover a wide
+  // dynamic range of numeric values, its precision not high enough to represent
+  // int64_t, so we need to be pay more attention when checking if a double is
+  // std::in_range<int64_t>().
+  static_assert(sizeof(double) == sizeof(int64_t));
+  auto int64_type = make_shared<long_type_impl>();
+
+  auto closest_to_max = IEEE754_double::closest(std::numeric_limits<int64_t>::max());
+  // from_json_object() should not throw at seeing: std::numeric_limits<int64_t>::max() - 1
+  //
+  // well sort of, as the 64 bit float point is not to present this number, we pick the
+  // last greatest 64 bit float point number less than std::numeric_limits<int64_t>::max() .
+  rjson::value less_than_int64_t_max{closest_to_max.previous()};
+  BOOST_CHECK_NO_THROW(from_json_object(*int64_type, less_than_int64_t_max));
+  // we don't test std::numeric_limits<int64_t>::max() here. because 64 bit IEEE-754 float
+  // point is not able to represent std::numeric_limits<int64_t>::max(),
+  // i.e., 9223372036854775807. if we cast it to double, we'll have a closest float number,
+  // i.e., 9223372036854775808.
```

Comment:
Doesn't this mean that the same bug you tried to fix still exists? (this is exactly the example from https://github.com/scylladb/scylladb/issues/10137).